File: CIS/matlab_scripts/gethr.m
function [hrp,hr]=gethr(sig,fs,f0)

% % % inputs:
% % % sig: recorded signal
% % % fs: sampling freq in hz
% % % f0: a rough estimate of what the pitch freq (hz) can be
% % % 
% % % Outputs:
% % % hrp: heart rate period in seconds
% % % hr: heart rate in hz

sig_o=sig; 
sig(abs(sig)<0.8*max(abs(sig)))=0;

crit_ind=length(sig)+round(0.5455*fs); %corr to 110bpm 60/110=.5455
% crit_ind=length(sig)+round(0.05*fs); %corr to 110bpm 60/110=.5455
c_sig=xcorr(sig);
F0=round(fs/f0);
F0_range=(F0-round(F0/2)):(F0+round(F0/2));
tmpind=length(sig)+F0_range;
if tmpind(1)<crit_ind
    ids=find(tmpind>=crit_ind);
    tmpind=tmpind(ids);
else
    tmpind=[crit_ind:tmpind(1)-1 tmpind];
    
end
if tmpind(end)>length(c_sig)
    ids=find(tmpind<=length(c_sig));
    tmpind=tmpind(ids);
end
% [tmpind(1) tmpind(end) length(c_sig)]
tmp=c_sig(tmpind);
% subplot(211);plot(sig_o);hold on;plot(sig,'r');hold off;
% subplot(212);plot(c_sig);hold on;plot(tmpind,tmp,'r');hold off;
[val ind]=max(tmp);
P0=tmpind(ind)-length(sig);
hr=1/(P0/fs);
hrp=1/hr;

File: CIS/utils/gethr.py
import numpy as np
import math

def gethr(sig, fs,f0):
    fs = 2000
    sig_o = (sig-np.mean(sig))/(np.max(sig)-np.min(sig))
    sig_o[np.argwhere(np.abs(sig_o)<0.2*np.max(np.abs(sig_o)))] = 0
    crit_ind = len(sig_o) + np.round(0.5455*fs)
    c_sig = np.correlate(sig_o,sig_o,mode = "full") #[len(sig_o)-1:]
    F0 = np.round(fs/f0)
    F0_range = np.arange((F0-np.round(F0/2)),(F0+np.round(F0/2)+1))
    tmpind = len(sig)+F0_range
    if tmpind[0] <crit_ind:
        ids = np.argwhere(tmpind>=crit_ind)
        tmpind = tmpind[ids].reshape(-1)
    else:
        tmpind = np.concatenate((np.arange(crit_ind,tmpind[0]-1), tmpind))

    if tmpind[-1]>len(c_sig):
        ids = np.argwhere(tmpind<=len(c_sig))
        tmpind = tmpind[ids].reshape(-1)

    tmp = c_sig[tmpind.astype(np.int64)]
    val = np.max(tmp)
    ind = np.argmax(tmp)
    P0 = tmpind[ind]-len(sig)
    hr = 1/(P0/fs)
    hrp = 1/hr
    
    return hrp,hr
